### ISSUE TITLE
Properly check for use-after-move and (maybe) use-after-drop

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#188]
Miri currently considers `Copy` and `Move` the same operation. There are some proposals in particular by @eddyb to make them not the same: it might be beneficial for optimizations to be able to rely on the data not being read again after a `Move`. This would correspond to replacing the data by `Undef`. It seems reasonable to do similar things for drop -- though that requires a precise definition of what is actually considered a drop (does that include just the `Drop` terminator or also calls to `drop_in_place`?).

One reason why I am a bit hesitant about this is that this makes read have a side-effect, so we'd have to make all read methods in Miri take the interpreter context by `&mut self`. (Reads already have a side-effect in Stacked Borrows but that is fairly local and we just use a `RefCell` there.

@eddyb 	what would be such optimizations that benefit from this, where `StorageDead` is not happening so we need to rely on `Move`?

**Comment by RalfJung** [comment]
I moved this to the UCG repo because really this is a question of what the semantics should be, not how to implement it in Miri.

**Comment by RalfJung** [comment]
Jim Blandy asked via email if similar things to what has been proposed for `Move`, apply to `ptr::read`: should that leave the memory from which things came as `Undef`? I think for `Copy` times the answer is definitely "no", but other types, if we decide to make `Move` "kill" the memory it came from, we might or might not want to do the same for `ptr::read`.

**Comment by RalfJung** [comment]
Regarding use-after-drop: as @CAD97 pointed out, `rc::Weak` is making the assumption that `size_of_val_raw` can be called on some value after it has been `drop_in_place`d. If we ever have thin DSTs with a vtable in the pointee, that would entail reading the vtable pointer after the object has been dropped.

**Comment by RalfJung** [comment]
See https://github.com/rust-lang/rust/issues/71117#issuecomment-864592461 for a possible reason why we might need moves to deinit their source.

**Comment by RalfJung** [comment]
On the other hand, it looks like we actually need to be able to read enum discriminants after partial moves out of an enum (https://github.com/rust-lang/rust/issues/91029, https://github.com/rust-lang/rust/pull/89764#issuecomment-973820415). That is a problem if that move de-inited the part of the enum where the discriminant is stored -- can that happen with niche optimizations?

**Comment by Lokathor** [comment]
An example would probably be Option<Vec<T>>. Vec is a ptr/len/capacity but the ptr is always a non-null so that makes a niche where the `None` part of `Option<Vec<T>>` goes. I'm not sure if that entirely fits the scenario you were imagining.

**Comment by RalfJung** [comment]
Well, the motivation for 'de-init on move' came up in https://github.com/rust-lang/rust/issues/71117#issuecomment-864592461, so the question is if that would ever be in conflict with something like `Option<Vec>`.

**Comment by Lokathor** [comment]
Ah, i think that part of it is a little too theoretical for me.

(though "some kind of aliasing rules" like you suggest at the end of your comment is often how I've imagined it to work)

**Comment by JakobDegen** [comment]
I can't find it now, but I vaguely remember that @bjorn3 mentioned this having use in codegen. Please let me know if I'm wrong, but my vague recollection is that if we have a parameter to a function that cannot be passed in registers, but rather must be passed via pointer, most calling conventions allow the callee to take ownership of the stack slot associated with the pointer. This means that if it was legal to use a value after moving it into another function, we would have to duplicate that value on the stack first.

**Comment by JakobDegen** [comment]
Since originally writing that comment, I've thought about this a bunch more, and am this point strongly of the opinion that we should deinit on move, and that furthermore moves should be a "write" in SB terms. For a concrete example of an optimization this enables, consider:

```rust
#[derive(Debug)]
pub struct NonCopy(i32);

fn foo(mut a: NonCopy, p: *const NonCopy) -> NonCopy {
    a = NonCopy(1);
    unsafe { p.read() }
}

pub fn bar() -> NonCopy {
    let s: NonCopy = NonCopy(0);
    let p = &s as *const _;
    foo(s, p)
}

fn main() {
    dbg!(bar());
}
```

Codegen would like to pass `s` in `bar` to `foo` by reference instead of generating a copy. This would be justified by the move of `s` invalidating all pointers to it and deinitializing it. That means the above code is UB. As a matter of fact, codegen does this optimization today (and the only reason we do not miscompile the above program is that MIR building generates an extra copy). I recently [did a perf run](https://github.com/rust-lang/rust/pull/96153) of the cost of disabling this, and the results were significant (although not on their own enough to justify a decision either way here).

I'm going to attempt to implement this in Miri/CTFE at some point (although might not get around to it for a bit), so that we can assess the impact of this change.

There was a comment in a related issue talking about the behavior of `ptr::read`. To be clear, I am by no means encouraging extending this behavior to `ptr::read`, `ptr::copy`, `ptr::copy_nonoverlapping` or any such functions. Turning these operations into an SB write of the source is a complete non-starter imo, as I expect this to break huge amounts of code. However, I do not think these two goals are conflicting. We can have this behavior only on those cases that are moves from the pov of borrowck (ie `Operand::Move` in MIR)

**Comment by CAD97** [comment]
As an example where pointer reads deiniting the place being bad (beyond just breaking the documented guarantees of duplicating the value), I just want to drop this example I thought of:

```rust
let mut x = Some(ManuallyDrop(NonZeroU8::new(1)));
match x {
    Some(md) => MaybeUninit::take(&mut md),
    None => 0,
};
x.is_some()
```

`ManuallyDrop::take` can't deinit the place it moves from because it could be used as a niche. (And unlike `MaybeUnunit`, `ManuallyDrop` preserves niches.)

**Comment by CAD97** [comment]
> If we ever have thin DSTs with a vtable in the pointee, that would entail reading the vtable pointer after the object has been dropped.

[ThinBox](https://doc.rust-lang.org/nightly/std/boxed/struct.ThinBox.html) exists on nightly now that does this for uniquely owned places, though it can't yet be done for `Rc`.

**Comment by RalfJung** [comment]
> I'm going to attempt to implement this in Miri/CTFE at some point (although might not get around to it for a bit), so that we can assess the impact of this change.

Sounds good! I am definitely in favor of this (including making a move count like a Stacked Borrows write, i.e., invalidating other pointers).

**Comment by digama0** [comment]
Could you give an example? If a move is also a last use then I don't see how the destructive move effect is observable.

**Comment by RalfJung** [comment]
If a move was always the last use then we wouldn't need the operational semantics to deinit them. ;)

If we consider `move(local)` and `copy(*ptr)`, then we cannot reorder them because `ptr` might point to `local`.

**Comment by RalfJung** [comment]
> Is there a reason not to implement it, other than implementation-level difficulties?

Yes. It makes MIR a lot harder to analyze, since the order in which operands are evaluated suddenly matters.

Also IMO this is the wrong question: when it comes to adding UB, we should always be asking whether there is a good reason to have this be UB, and make it not UB otherwise.

**Comment by afetisov** [comment]
Whether it is technically UB or not doesn't matter. The basic rules of Rust are that moves are destructive, so the old location must never be used again. Even if it's not UB, reading the moved-from place would still be a major logical error, because that moved object can have all kinds of external resources attached or reasons to be non-cloneable. I expect tools like Miri to help in finding double drops, regardless whether it is UB in the technical sense which could affect generated code.

> It makes MIR a lot harder to analyze, since the order in which operands are evaluated suddenly matters.

That may be true, but since moves must be destructive, you should still encode the invalidation of old places in some way, If the question is "do we invalidate on move, or do we encode it is a separate operation", then I have no opinion which implementation strategy is better, but if the question is "do we keep moved-from places initialized and usable", then it violates the expected semantics of moves and drops.

**Comment by RalfJung** [comment]
This repository, and Miri, are all about whether code is "technically" UB or not. That is literally the only question that matters here.

**Comment by Zeegomo** [comment]
Looks like this discussion mostly revolves around `move`s, while maybe the thing for `drops` is a little bit easier (and some recent proposals try to decouple those).
For types with destructor it seems to me that reading the value again should be undefined behavior (for example because the drop implementation violates the, possibly custom, guarantees of the type, e.g. by releasing the backing memory of a `Vec`).


Not sure if it's helpful, but I have this weird idea that a drop (both `TerminatorKind::drop` and `drop_in_place`, as they both execute the destructor) might turn the type of the dropped value from `T` to  a special `Dropped([u8; size_of::<T>()])`. 
This would release all type invariants on the value, while still preserving the raw bytes and size of the type (not sure there's a way for the destructor to mess with that)

**Comment by JakobDegen** [comment]
@Zeegomo I don't see why that means you need any amount of UB - using a value of an unknown type after it has been dropped is clearly unsound, there's nothing that making it UB gets you.

**Comment by Zeegomo** [comment]
Not sure what you mean by unknown type. 
The caller of `drop_in_place` might call it on a vec and then try to access it again. That subsequent use of vec should be UB.
Have I misunderstood something?

**Comment by CAD97** [comment]
The difference between library UB and language UB. Using a value after it's been dropped can obviously cause UB. But that's library UB, the same as using `ptr::copy` to duplicate a value and using both can cause UB.

Library UB is when you've misused an `unsafe` API and now the use of safe functions could cause language UB (or using a different version of the library, or of course immediately, etc); you've violated an API requirement and the library no longer guarantees freedom from language UB.

Library UB doesn't mean that the program execution is necessarily undefined, though. As a simple example, consider that a possible implementation of any unsafe `new_unchecked` is to just call the safe `new`, which obviously has defined behavior. Language UB is an absolute terror; the result is *mathematically undefined*, the compiler assumes that it never happens, and as such if UB actually does happen then *retroactively* all guarantees about your program are voided and your debugger has been lying to you since the beginning.

By saying that reading memory after it's been dropped is UB, you're saying that the example program of

```rust
let p: *mut Box<i32> = Box::into_raw(Box::new(Box::new(0)));
drop_in_place(p);
let x: *mut i32 = *(p as *mut *mut i32);
dbg!(x);
```

should be UB and that there is no incorrect compilation of this program. The semantically very similar

```rust
let p: *mut Box<i32> = Box::into_raw(Box::new(Box::new(0)));
drop(ptr::copy(p));
let x: *mut i32 = *(p as *mut *mut i32);
dbg!(x);
```

is unambiguously a defined program.

The extra fun version is that by saying reading the memory is UB, that's an even stronger requirement than saying the memory is uninitialized. Abstract bytes now have a 258th state (still ignoring provenance) where the mere act of reading them even as `MaybeUninit<u8>` is UB.

Personally, my current leaning is that running drop glue on a place (via `drop_in_place` or otherwise) should be defined to leave the bytes in the state that the `Drop::drop` implementation(s) leave(s) them. This is [potentially necessary](https://github.com/rust-lang/lang-team/blob/master/src/design_notes/dynsized_constraints.md) for `Rc<T>` for certain new kinds of dynamically sized `T`. (`Rc`'s current implementation relies on the fact that it's sufficient for validity to call [`Layout::for_value_raw`](https://doc.rust-lang.org/stable/std/alloc/struct.Layout.html#method.for_value_raw) on a pointer to dropped value.

I'm also of the personal opinion that a move should be properly destructive and uninitialize the memory being moved from. (Probably in separate typed copy and uninitialize steps at the opsem level, but imo ideally in a way such that the two places are allowed to have the same address if they're both stack allocated objects.)

**Comment by Zeegomo** [comment]
>By saying that reading memory after it's been dropped is UB,

I might have used the wrong terminology, but that's not what I meant to say. The destructor might leave some fixed byte string in place of the value, but that fixed value has no special meaning besides being a byte string (i.e. the destructor can leave the value in a state that violates type guarantees).  Using that value again as a value of any type T requires great care   and possibly cause UB, as would `transmute::<[u8], T>()`

**Comment by digama0** [comment]
> The destructor might leave some fixed byte string in place of the value, but that fixed value has no special meaning besides being a byte string (i.e. is the destructor allowed to leave the value in a state that violates type guarantees?).

This is the case for *any* unsafe function which can obtain a pointer to the allocation. It is not language UB to write `2` to a stack variable of type `bool`, it is only UB if you perform a *typed copy* of that invalid value at the original type, for example by returning it from a function or reading the variable after it was trashed. If you just let it go out of scope then this is DB (since `bool` doesn't have a destructor).

**Comment by RalfJung** [comment]
I think there is a somewhat separate question here. Even if `b` is guaranteed to have storage until it goes out of scope, should its 'type' really matter after it has been dropped? Arguably it is just raw storage at this point, and could be considered `Send`.

Note that this is not even primarily a question of motivation. Having storage disappear earlier is hard, I haven't even seen a proposal that actually achieves this. We'd need to emit suitable MIR statements indicating that `b` is deallocated, so that Miri can flag appropriate UB. It is certainly not sufficient to just change the generator lowering, we need the following code to be UB (flagged by Miri):
```rust
struct B(i32);

fn main() {
    let mut b = B(0);
    let ptr = std::ptr::addr_of_mut!(b);
    drop(b);
    unsafe { ptr.write(B(1)); }
}
```

**Comment by saethlin** [comment]
This sound to me like tightening up the way we handle `StorageLive`/`StorageDead`. Or just revisiting that whole system. Selfishly, I would really like to see that area of MIR semantics revisited because it is incredibly hard to work with in MIR optimizations.

**Comment by nbdd0121** [comment]
> I think there is a somewhat separate question here. Even if `b` is guaranteed to have storage until it goes out of scope, should its 'type' really matter after it has been dropped? Arguably it is just raw storage at this point, and could be considered `Send`.

Yeah that could be one way to tackle the `Send` problem. However we still need to make `b`'s storage dead in order to reduce the size of the generator.

**Comment by RalfJung** [comment]
I created a new issue specifically for "what is the semantics of `move` (in current codegen, and beyond)": https://github.com/rust-lang/unsafe-code-guidelines/issues/416.

**Comment by JakobDegen** [comment]
Visiting for backlog bonanza. This may need deduplication with some other open issues, but leaving open as the canonical one

**Comment by RalfJung** [comment]
Above and elsewhere, I said things like

> making a move count like a Stacked Borrows write, i.e., invalidating other pointers

I actually today realized that while this is true for locals, it is not true for places inside a `Box` or `&mut`. Consider:
```rust
mod other {
    /// Some private memory to store stuff in.
    static mut S: *mut i32 = 0 as *mut i32;

    pub fn lib1(x: & &mut i32) { unsafe {
        S = (x as *const &mut i32).cast::<*mut i32>().read();
    } }
    
    pub fn lib2() { unsafe {
        *S = 1337;
    } }
}

fn main() {
    let x = &mut 0;
    other::lib1(&x);
    *x = 42; // a write to x -- invalidates other pointers?
    other::lib2();
    assert_eq!(*x, 1337); // oops, the value changed
}
```
Giving away `&x` lets the other module take a copy of the tag of `x`, which can later be used.

So at least for `move(*place)`, we don't get a guarantee that all other pointers have been invalidated.

And with Tree Borrows, even for `move(local)`, we don't get such a guarantee. This is because Tree Borrows does not retag on `addr_of_mut!(local)`. The fact that Stacked Borrows retags here causes tons of problems so I strongly prefer the Tree Borrows semantics.

---

I also realized I didn't reply to this part by @arielb1 
> And as for address inequality, I'm for that if a local is currently deinitialized (i.e. trying to borrow it would give you a "use of moved value" error), then it is allowed for other locals to have the same address as it. 

If a local has storage then it must have a distinct address, even if that storage is uninitialized.

The way to achieve potential equal addresses is to remove its storage. We already have a way to express that: `StorageDead`. I'm all for having HIR-to-MIR lowering move `StorageDead` earlier, and we need to carefully specify how early. But having a `move` be `StorageDead` makes little sense since `move` works on arbitrary places, such as `local.field`, and those cannot in general be deallocated.

So basically, replace the `b = poison` in your examples with `StorageDead(b)` and I agree. My position is still that any kind of use-after-move/drop UB should be fully explicitly encoded by deallocating the backing memory.

Now, you bring up rustc performance. Given that we have explicit `StorageDead` for many locals already, I wouldn't expect that moving that to earlier (namely when the move occurs, instead of when the local goes out of scope) would make MIR so much bigger.
